### DISCUSSION TITLE
How to use re-run service in webdriverio?

### DISCUSSION THREAD

**Discussion by yiyef** [webdriverio/webdriverio#6878] [Technical Discussion]
I read from this doc: https://webdriver.io/docs/wdio-rerun-service/, that the rerun service is what I need to rerun the whole process when a test fails. I installed the rerun service and insert this in my code

const RerunService = require('wdio-rerun-service');
export.config = {
    // ...
    services: [RerunService, {}],
    // ...
}; 
but I can not find out how to insert the rerun in my testing files, would sb help me with this? Thank you so much

**Comment by nimeshbhatt** [comment]
Hi @yiyef did you get how to use re-run service
or
Hi @klamping, could you please help how exactly we can use the Re-run service?

I have webdriverio project for automating web app in mobile browser, and I have used Appium, Jasmin and JavaScript as scripting language. 
this service has tracked the failed scenario by creating a JSON file but, I want to rerun that failed scenario. would you please help me to run the failed test cases.

**Comment by christian-bromann** [comment]
Maybe @esaari can elaborate on this?

**Comment by esaari** [comment]
The rerun will not happen automatically. A `rerun.sh` file will be generated and you'll need to set a rerun script in your `package.json` that will simply execute the `rerun.sh` script. If this is ambiguous, we should update our documentation.

**Reply by nimeshbhatt** [reply]
Hi @esaari , thanks for the reply. 

How exactly the  `rerun.sh` file will get generated? I added this service `rerunScriptPath: './path-to-generate-this-file/rerun.sh'` to generate the .sh file. But, it has not. I'll share you how I have wrote the service in my config file.

```
services: [
        'appium',
        [RerunService, 
            { 
                // commandPrefix: 'VARIABLE=true', 
                // ignoredTags: ['@known_bug'], 
                rerunDataDir: './conf/', 
                rerunScriptPath: './conf/rerun.sh'
            }
        ], 
    ],
```

By the above setting in config file, it is generating the .json file that is tracking the failed test cases but not generating the `rerun.sh` file.

**Reply by nimeshbhatt** [reply]
Hi @esaari, is there any update on the above mentioned issue? 

By the above setting in config file, it is generating the .json file that is tracking the failed test cases but not generating the `rerun.sh` file.

**Reply by poonamdce** [reply]
Hi @nimeshbhatt @esaari I am trying to use this package but exactly the same issue that rerun script is not generated automatically. Can you please share how did you make it work? It is urgent :(

**Reply by Emmet-McGuire** [reply]
> The rerun will not happen automatically. A `rerun.sh` file will be generated and you'll need to set a rerun script in your `package.json` that will simply execute the `rerun.sh` script. If this is ambiguous, we should update our documentation.

Could you please update the documentation to include what to add to your package.json to get the re-run to actually re-run?  It's not clear from the documentation that this is required, and I just hours debugging before finally finding this thread.  Thanks in advance.

**Reply by blockqa** [reply]
Hi @esaari Can you please update the docs, on how we can run rerun.sh file in package.json

**Reply by blockqa** [reply]
I can generate rerun.sh and I tried to run this file in  package.json, but it is not running successfully. 

-MacBook-Pro wdio-cucumber-e2e-test % npm run rerun                

> wdio-cucumber-e2e-test@1.0.0 rerun
> sh rerun.sh

2022-09-30T20:26:35.225Z WARN @wdio/config:ConfigParser: pattern rerun.sh did not match any file

Execution of 5 workers started at 2022-09-30T20:26:35.231Z

(node:23265) [DEP0147] DeprecationWarning: In future versions of Node.js, fs.rmdir(path, { recursive: true }) will be removed. Use fs.rm(path, { recursive: true }) instead
(Use `node --trace-deprecation ...` to show where the warning was created)

Spec Files:      0 passed, 5 skipped, 5 total (100% completed) in 00:00:04

**Comment by kerkeniakov** [comment]
Can't you just use specFileRetries in wdio.conf? I think this package is obsolete now.
e.g add specFileRetries: 2 to wdio.conf

**Reply by Emmet-McGuire** [reply]
@kerkeniakov Yes, I could but the behavior is slightly different.  I am using cucumber, and my understanding is that [specFileRetries](https://webdriver.io/docs/retry/) will only re-run either the whole feature file, or a single step (not a scenario).  Retrying a single step will most likely fail as it needs the setup of the previous steps to pass, and re-running the whole feature file is time prohibitive. 

This is the value of re-run service over using specFileRetries.

fwiw, I was able to get it running using a bash script that kicks off the initial test, and conditional logic to run the retry script if the initial runs fail.  I believe this is the intended way to handle it, but if re-run service is still supported, the official documentation should indicate this so other people know what to do.  I think most people think it's "plug and play" but it's not (specFileRetries is though).

**Comment by gunsingh-clgx** [comment]
Hi , 
After updating it to the lates version 2.0.3 , I am unable to generate rerun.json and rerun.sh files, can someone please help me out

**Reply by suleytron** [reply]
> Hi , After updating it to the lates version 2.0.3 , I am unable to generate rerun.json and rerun.sh files, can someone please help me out

I have the same issue. I was getting the .json file with wdio-rerun-serice 1.7.6 but no .rerun.sh file. I upgraded to wdio-rerun-service 2.0.3 and now I don't get anything. Did anyone manage to resolve this?

**Reply by esaari** [reply]
Sorry, I'm just noticing this discussion. This issue should be resolved with the latest `2.0.6` release.

**Reply by firmansyahrosid** [reply]
hi @esaari I used the latest 2.0.7 version, but rerun.sh file is still not generated.

**Comment by esaari** [comment]
FYI, I enabled the discussions board within the rerun project itself: https://github.com/webdriverio-community/wdio-rerun-service/discussions

I kindly request that any current or future discussions happen there, as it will be easier for the maintainers to engage and provide help. 
(cc @mikesalvia)
